File: Leetcode/16.最接近的三数之和.cpp
/*
 * @lc app=leetcode.cn id=16 lang=cpp
 *
 * [16] 最接近的三数之和
*/

#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

/*2021-09-30-10-34*/

//from https://github.com/fengwei2002/Algorithm
/* 
 !  注意根据距离更新 ans 的写法，剩下的代码 和三数之和一模一样
 */

// @lc code=start
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        int n = nums.size();
        int ans = nums[0] + nums[1] + nums[2];

        for (int i = 0; i < n; i++) {
            if (i - 1 >= 0 && nums[i] == nums[i - 1]) continue; 

            int k = n - 1;
            int j = i + 1;

            while (j < k) {
                // 注意这三行根据距离更新 ans 的方式
                int dummy = nums[i] + nums[j] + nums[k];
                if (abs(dummy - target) < abs(ans - target))
                    ans = nums[i] + nums[j] + nums[k];

                if (dummy < target) {
                    j++;
                } else if (dummy > target) {
                    k--;
                } else {
                    while (j < k && nums[k] == nums[k - 1]) k--;
                    while (j < k && nums[j] == nums[j + 1]) j++;
                    j++, k--;
                }
            }
        }
        return ans;
    }
};
// @lc code=end

File: Leetcode/16.最接近的三数之和.go
/*
 * @lc app=leetcode.cn id=16 lang=golang
 *
 * [16] 最接近的三数之和
 */
package leetcode

/* --- 2022-04-06-21-38 --- */
// https://github.com/fengwei2002/Algorithm

import (
	"sort"
	"math"
)

// @lc code=start
func threeSumClosest(nums []int, target int) int {
	n := len(nums)
	ans := 0
	diff := math.MaxInt32
	sort.Slice(nums, func(i, j int) bool {
		return nums[i] < nums[j]
	}) // 要先排序的话，才具有双指针的单调性

	for i := 0; i < n; i++ {
		for j, k := i + 1, n - 1; j < k; j++ {
			for j < k - 1 && nums[i] + nums[j] + nums[k - 1] >= target {
				k--
			}
			// 此时，nums[i, j, k] 和 nums[i, j, k - 1] 就是 target 边界的左右值 
			sum := nums[i] + nums[j] + nums[k]
			newDiff := abs(sum - target)
			if newDiff < diff {
				ans = sum
				diff = newDiff
			}
			if k - 1 > j {
				sum = nums[i] + nums[j] + nums[k - 1]
				newDiff = abs(sum - target)
				if newDiff < diff {
					ans = sum
					diff = newDiff
				}
			}
		}
	}
	return ans
}

func abs(a int) int { 
	if a < 0 {
		return -a
	} 
	return a
}
// @lc code=end

